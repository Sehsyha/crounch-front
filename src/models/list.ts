import { Product } from './product';
import { CategoryInGetListResponse } from './category';

interface List {
  id: string
  name: string
  products: Product[]
}

interface GetListResponse {
  id: string
  name: string
  creationDate: string
  categories: CategoryInGetListResponse[]
}

export { List, GetListResponse }
