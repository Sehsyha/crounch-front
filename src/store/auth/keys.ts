export enum AuthActions {
  LOGIN = 'LOGIN',
  SIGNUP = 'SIGNUP',
  LOGOUT = 'LOGOUT',
}

export enum AuthGetters {
  IS_AUTHENTICATED = 'IS_AUTHENTICATED',

}

export enum AuthMutations {
  LOGIN = 'LOGIN',
  LOGOUT = 'LOGOUT',
}
