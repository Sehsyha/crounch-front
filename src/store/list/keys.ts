export enum ListKeys {
  CREATE = 'CREATE',
  CREATED = 'CREATED',
  GETOWNERS = 'GETOWNERS',

  ADD = 'ADD',
  SET = 'SET',
  RESET = 'RESET',
};
