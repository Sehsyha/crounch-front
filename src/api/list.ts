import { doFetch } from './api';
import { List } from '@/models/list';

export function createList(name: string): Promise<List> {
  return doFetch({
    url: 'lists',
    method: 'POST',
    data: { name }
  });
}

export function getOwnerLists(): Promise<Array<List>> {
  return doFetch({
    url: 'lists',
    method: 'GET',
  });
}

export function deleteList(id: string): Promise<void> {
  return doFetch({
    url: `lists/${id}`,
    method: 'DELETE',
  });
}

export function addProductToList(productID: string, listID: string): Promise<void> {
  return doFetch({
    url: `lists/${listID}/products/${productID}`,
    method: 'POST'
  })
}

export function readList(listID: string): Promise<List> {
  return doFetch({
    url: `lists/${listID}`,
    method: 'GET'
  })
}
