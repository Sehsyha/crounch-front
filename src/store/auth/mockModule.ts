import { AuthKeys } from './keys';
import { Module } from 'vuex';
import { AuthState } from '.';
import { RootState } from '..';

const actions = {
  [AuthKeys.SIGNUP]: jest.fn(),
  [AuthKeys.LOGIN]: jest.fn(),
  [AuthKeys.LOGOUT]: jest.fn(),
};

const getters = {
  isAuthenticated: jest.fn(),
};

export function createAuthModuleMock(): Module<AuthState, RootState> {
  return {
    namespaced: true,
    actions,
    getters,
    state: {
      token: '',
      status: '',
    },
  };
}
