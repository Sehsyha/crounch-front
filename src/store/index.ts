import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import { auth } from './auth';
import { ListModule } from './ListModule';
import { product } from './product'

Vue.use(Vuex);

export interface RootState { }

const store: StoreOptions<RootState> = {
  state: {},
  modules: {
    auth,
    ListModule,
    product,
  },
  strict: true,
};

export default new Vuex.Store<RootState>(store);
