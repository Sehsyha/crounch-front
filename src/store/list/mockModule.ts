import { Module } from 'vuex';

import { ListKeys } from './keys';
import { ListState } from '.';
import { RootState } from '..';

const actions = {
  [ListKeys.CREATE]: jest.fn(),
  [ListKeys.CREATED]: jest.fn(),
  [ListKeys.GETOWNERS]: jest.fn(),
};

const getters = {
  lists: jest.fn(),
};

const mutations = {
  [ListKeys.RESET]: jest.fn(),
};

export function createListModuleMock(): Module<ListState, RootState> {
  return {
    namespaced: true,
    actions,
    getters,
    mutations,
    state: {
      lists: [],
    },
  };
}
