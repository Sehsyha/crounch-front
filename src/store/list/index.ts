import { Module } from 'vuex';

import { actions } from './actions';
import { RootState } from '..';
import { List } from '@/models/list';

export interface ListState {
  lists: List[];
}

export const state: ListState = {
  lists: [],
};

const namespaced: boolean = true;
export const listNamespace = { namespace: 'list' };

export const list: Module<ListState, RootState> = {
  namespaced,
  state,
  actions,
};
