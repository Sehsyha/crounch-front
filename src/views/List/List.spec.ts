import { Wrapper } from '@vue/test-utils';

import { shallowComponent } from '@/utils/test';
import ListPage from './List.vue';
import { Module } from 'vuex';
import { ListState } from '@/store/list';
import { ListGetters, ListActions } from '@/store/list/keys';
import { RootState } from '@/store';
import { createListModuleMock } from '@/store/list/mockModule';
import { List } from '@/models/list';

describe('List', () => {
  let wrapper: Wrapper<ListPage>;
  let listModule: Module<ListState, RootState>;
  const list: List = {
    id: 'list id',
    name: 'list name',
  };

  beforeEach(() => {
    listModule = createListModuleMock();
    (listModule.getters![ListGetters.GET] as jest.Mock).mockReturnValue(() => list);
    shallowListComponent();
  });

  it('Should render.', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Should not fetch lists when list is defined.', () => {
    shallowListComponent();

    expect(listModule.actions![ListActions.GETOWNERS]).not.toHaveBeenCalled();
  });

  it('Should fetch lists when list is undefined.', () => {
    (listModule.getters![ListGetters.GET] as jest.Mock).mockReturnValue(() => undefined);

    shallowListComponent();

    expect(listModule.actions![ListActions.GETOWNERS]).toHaveBeenCalled();
  });

  it('Should display list when the list is defined after fetch.', done => {
    let calledOneTime = false;
    (listModule.getters![ListGetters.GET] as jest.Mock)
      .mockImplementation(() => () => new Promise(resolve => {
        if (!calledOneTime) {
          calledOneTime = true;
          return Promise.resolve(undefined);
        }
        return Promise.resolve(list);
      }));

    shallowListComponent();

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  it('Should display not found message when list does not exist after fetch.', done => {
    (listModule.getters![ListGetters.GET] as jest.Mock).mockReturnValue(() => undefined)
    shallowListComponent();

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot()
      done();
    });
  });

  function shallowListComponent() {
    const modules = {
      list: listModule,
    };

    const values = {
      id: list.id,
    };

    wrapper = shallowComponent(ListPage, { modules, values });
  }
});
