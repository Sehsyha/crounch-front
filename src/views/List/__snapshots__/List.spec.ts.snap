// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Should display list when the list is defined after fetch. 1`] = `
<div>
  <div></div>
</div>
`;

exports[`List Should display not found message when list does not exist after fetch. 1`] = `
<div>
  <div>La liste n'a pas été trouvée</div>
</div>
`;

exports[`List Should render. 1`] = `
<div>
  <div>list name</div>
</div>
`;
